### PR TITLE
 [clang][analyzer] Improve BlockInCriticalSectionsChecker

### PR DESCRIPTION
**Pull request by gamesh411**
On entering a critical section, a note tag is now placed along the bugpath.

### CODE REVIEW

**Review by NagyDonat — APPROVED** [review]
LGTM.

**Comment by balazske** [comment]
I think we could get a note at the first `lock` call in this code, this behavior is not entirely correct:
```
void test() {
  std::mutex m;
  m.lock();
  m.unlock();
  m.lock();
  sleep(3);
  m.unlock();
}
```

File: clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp [hunk 1]
```diff
@@ -57,6 +61,12 @@ class BlockInCriticalSectionChecker : public Checker<check::PostCall> {
                                 const CallEvent &call,
                                 CheckerContext &C) const;
 
+  CritSectionMarker getCriticalSectionMarker(const CallEvent &Call,
+                                             CheckerContext &C) const;
+  const NoteTag *
+  createCriticalSectionNote(const CritSectionMarker &CriticalSectionBegin,
+                            CheckerContext &C) const;
```
Comment (balazske) [inline clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:68]: I do not know if it is sufficient to store the set of all found lock calls until the last unlock happens. Probably in the following situation it does not work:
```
std::mutex m, n, k;
m.lock();
n.lock();
n.unlock();
k.lock();
k.unlock();
sleep(1);
m.unlock();
```
It looks better to store all locked mutex objects somehow (not the lock calls), and remove one at unlock. Then the mutex counter is not needed.

**Comment by gamesh411** [comment]
As keeping the trivial change that this PR was before, and thanks to the fact that the checker needed a major overhaul, I have increased the scope of this change to support the use cases mentioned by @balazske.
That is why I am re-requesting a review.

**Review by NagyDonat — COMMENTED** [review]
Seems to be a good change overall and I like that you implemented this with some nice helper classes. However, there is some code duplication that could be reduced (see inline comments).

Also, I feel that (especially after these extensions) this checker duplicates some logic that also appears in `PthreadLockChecker.cpp`. I didn't check the exact details but that checker is also modeling the locking and unlocking calls in order to report bad locking/unlocking order.

In theory it would be to ensure that we don't develop and maintain two similar but different solutions that both track lock handling code -- but PthreadLockChecker is 700 lines long and I understand that it would be difficult to unify it with this checker.

File: clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp [hunk 2]
```diff
@@ -70,73 +202,121 @@ class BlockInCriticalSectionChecker : public Checker<check::PostCall> {
...
+      llvm::find_if(MutexDescriptors, [&Call](auto &&UnlockFn) {
+        return std::visit(
+            [&Call](auto &&Descriptor) {
+              return Descriptor.matchesUnlock(Call);
+            },
+            UnlockFn);
+      });
+  if (UnlockDescriptor != MutexDescriptors.end())
+    return *UnlockDescriptor;
+  return std::nullopt;
```
[the remark this answers is not in the document]
Comment (gamesh411) [reply clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp]: I think I have fixed most of the code duplication, thanks for the suggestions 👍

File: clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp [hunk 3]
```diff
@@ -20,48 +20,180 @@
...
+  }
+  [[nodiscard]] bool matchesUnlock(const CallEvent &Call) const {
+    initIdentifierInfo(Call);
+    const auto *Dtor = dyn_cast<CXXDestructorCall>(&Call);
+    if (!Dtor)
+      return false;
+    auto *IdentifierInfo =
+        cast<CXXRecordDecl>(Dtor->getDecl()->getParent())->getIdentifier();
+    return IdentifierInfo == Guard;
+  }
```
[the remark this answers is not in the document]
Comment (gamesh411) [reply clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp]: Fixed.

File: clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp [hunk 4]
```diff
@@ -20,48 +20,180 @@
...
+  bool matchesUnlock(const CallEvent &Call) const {
+    return UnlockFn.matches(Call);
+  }
+  [[nodiscard]] const MemRegion *getLockRegion(const CallEvent &Call) const {
+    return cast<CXXMemberCall>(Call).getCXXThisVal().getAsRegion();
+  }
+  [[nodiscard]] const MemRegion *getUnlockRegion(const CallEvent &Call) const {
+    return cast<CXXMemberCall>(Call).getCXXThisVal().getAsRegion();
+  }
+};
```
[the remark this answers is not in the document]
Comment (gamesh411) [reply clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp]: Introduced a common base class for these 2 classes.

File: clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp [hunk 5]
```diff
@@ -20,48 +20,178 @@
...
+  CallDescription UnlockFn;
+
+public:
+  CallDescriptionBasedMatcher(CallDescription &&LockFn,
+                              CallDescription &&UnlockFn)
+      : LockFn(std::move(LockFn)), UnlockFn(std::move(UnlockFn)) {}
+  [[nodiscard]] bool matches(const CallEvent &Call, bool IsLock) const {
+    if (IsLock) {
+      return LockFn.matches(Call);
+    }
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:69]:
```suggestion
    if (IsLock)
      return LockFn.matches(Call);
```
Bikeshedding: the coding standard says that braces shouldn't be used on simple single-statement bodies: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

Unorthodox but (IMO) elegant alternative idea: replace the whole body of the function with
```
return (isLock ? LockFn : UnlockFn).matches(Call);
```

File: clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp [hunk 6]
```diff
@@ -20,48 +20,178 @@
...
+    return II == Guard;
+  }
+
+public:
+  RAIIMutexDescriptor(StringRef GuardName) : GuardName(GuardName) {}
+  [[nodiscard]] bool matches(const CallEvent &Call, bool IsLock) const {
+    initIdentifierInfo(Call);
+    if (IsLock) {
+      return matchesImpl<CXXConstructorCall>(Call);
+    }
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:127]:
```suggestion
    if (IsLock)
      return matchesImpl<CXXConstructorCall>(Call);
```
Bikeshedding as above.

File: clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp [hunk 7]
```diff
@@ -20,48 +20,178 @@
...
+    return matchesImpl<CXXDestructorCall>(Call);
+  }
+  [[nodiscard]] const MemRegion *getRegion(const CallEvent &Call,
+                                           bool IsLock) const {
+    const MemRegion *LockRegion = nullptr;
+    if (IsLock) {
+      if (std::optional<SVal> Object = Call.getReturnValueUnderConstruction()) {
+        LockRegion = Object->getAsRegion();
+      }
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:136]:
```suggestion
      if (std::optional<SVal> Object = Call.getReturnValueUnderConstruction())
        LockRegion = Object->getAsRegion();
```
Yet another brace tweak...

**Review by NagyDonat — APPROVED** [review]
Thanks for simplifying the code :)

I added some stereotypical bikeshedding in inline comments, but apart from that the commit LGTM.

**Comment by balazske** [comment]
It looks like that this change causes crashes on many projects (curl, vim, postgres, others) in `RAIIMutexDescriptor::initIdentifierInfo`.

**Pull request merged**